### PR TITLE
Fixes JS and CSS errors in inline and course-wide Discussion XBlocks

### PR DESCRIPTION
**Pull request by pomegranited**
This change fixes a number of Javascript errors and warnings and ensures that the inline DiscussionXBlock, the DiscussionCourseXBlock, and the main Discussion Forum function as on edx:master.

**JIRA tickets**: Implements [YONK-401](https://openedx.atlassian.net/browse/YONK-401).

**Screenshots**:

Inline DiscussionXBlock:
inline discussionxblock

DiscussionCourseXBlock (staff view):
discussion course xblock - staff

DiscussionCourseXBlock (student view):
discussion course xblock - student

Discussion Forum:
discussion forum - student

**Testing instructions**:

1. Get edx-solutions running on the parent branch, `ziafazal/fix-broken-features`, then checkout this PR's branch `jill/discussion-xblock-bugfix`.
1. Restart Studio and LMS, allowing the static assets to rebuild (`paver devstack lms`, `paver devstack studio`, without `--fast`).
1. As `staff` user, create a course in Studio.
1. Under `Settings > Advanced Settings > Advanced Module List`, add `["discussion-course"]`.
1. Add an inline discussion component via the main "Discussion" component button.
1. Add a course-wide discussion component via "Advanced > Discussion Course".
1. Publish the course.
1. Open up a developer console to watch for Javascript warnings and errors during the following steps.
1. As `staff` user, load the Course Discussion unit.  Ensure that no discussion-related javascript errors or warnings are shown, namely no messages like:

    ```
Unable to load Sequence: DiscussionApp is not defined
backbone.js:1659: Uncaught Error: Backbone.history has already been started
jquery.js:5434: GET http://localhost:8000/c4x/OpenCraft/IDX101/asset/sass_discussion-forum.css fails
```
1. Visit the Inline Discussion unit, and ensure that clicking `Expand discussion` shows the current posts and an Add Post button, instead of an error modal dialog box.
1. As `staff` user, ensure that you can perform the following actions on your posts from within the Inline, Course Discussion, and main Discussion Forum tab:
    * Create Post
    * Reply to Post
    * Add Comment to a Reply
    * Edit a Post, Reply, and Comment.
    * Test [MathJax](http://docs.mathjax.org/en/latest/start.html) functionality in Post, Reply, Comment (and Preview), with e.g.

       ```
There are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$

When $a \ne 0$
```

    * View Discussion User Profile pages by clicking on the user's name.  Posts should render, including MathJax markup.
    * Delete a Post, Reply, and Comment.
    * Open/close a thread/post - in the Instructor panel, add the `staff` user to the `Discussion Admin` or `Discussion Moderator` group to get access to these functions.
    * Follow/unfollow a thread/post
    * Pin/unpin a thread/post
    * Search for posts by clicking the search button, or by hitting Enter in the search box.
    * With the LMS page loaded, stop the LMS service, and try to Create a Post.  Check that the error modal shows up.
1. Login as a student user and register for your course.  You should be able to do the above actions, plus:
    * Report/unreport a post created by the staff user.
    * Vote/unvote a thread/post created by the staff user.

**Author notes and concerns**:

1. Out of scope for this task is to move the code from [`lms.django_comments_client.forum`] to [`lms.discussion`](https://github.com/edx/edx-platform/tree/master/lms/djangoapps/discussion).  cf `edx:master` changes starting from [89f93df](https://github.com/edx/edx-platform/commit/89f93df46e6227058a4c7ef5c00cdb4a456e4583#diff-f0afdf1bdfbfc3a47f6722941269b4a4).
1. Fixing test failures is an ongoing process that is part of the overall rebase project.

**Reviewers**
- [ ] @itsjeyd 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

File: common/static/common/js/discussion/views/discussion_content_view.js [hunk 1]
```diff
@@ -366,41 +366,41 @@
...
+                        type: 'POST',
+                        data: {endorsed: is_endorsing},
+                        $elem: $(event.currentTarget)
                     },
-                    beforeSend: beforeFunc,
-                    $elem: $(event.currentTarget)
-                }, msg).always(this.trigger("comment:endorse"));
+                    msg,
+                    function() { return self.trigger('comment:endorse'); }
+                ).always(this.trigger('comment:endorse'));
```
Comment (itsjeyd) [inline common/static/common/js/discussion/views/discussion_content_view.js]: @pomegranited On this branch, `DiscussionUtil.updateWithUndo` only takes four arguments, see [utils.js#L234](https://github.com/open-craft/edx-platform/blob/131b1dda707e21b1fb26ec368b3458f88435e157/common/static/common/js/discussion/utils.js#L234). You're passing five here; is this causing any issues?

File: common/static/common/js/discussion/views/discussion_thread_profile_view.js [hunk 2]
```diff
@@ -42,31 +41,27 @@
...
                 return this;
             };
 
             DiscussionThreadProfileView.prototype.convertMath = function() {
-                return this.model.set(
-                    'markdownBody',
-                    DiscussionUtil.postMathJaxProcessor(DiscussionUtil.markdownWithHighlight(this.model.get('body')))
-                );
+                var htmlSnippet = DiscussionUtil.markdownWithHighlight(this.model.get('body'));
+                this.model.set('markdownBody', htmlSnippet);
```
Comment (itsjeyd) [inline common/static/common/js/discussion/views/discussion_thread_profile_view.js]: @pomegranited Just in case: OK to not use `DiscussionUtil.postMathJaxProcessor` here?

File: common/static/common/js/spec_helpers/discussion_spec_helper.js [hunk 3]
```diff
@@ -1,113 +1,121 @@
...
+                id: '567',
+                upvoted_ids: []
+            };
+        };
+
+        DiscussionSpecHelper.getTestRoleInfo = function() {
+            return {
+                Moderator: [],
+                Administrator: [],
+                'Community TA': []
```
Comment (itsjeyd) [inline common/static/common/js/spec_helpers/discussion_spec_helper.js]: @pomegranited `Moderator` and `Administrator` are not wrapped in quotes. Is that the case on `edx:master` as well?

File: common/static/common/js/spec_helpers/discussion_spec_helper.js [hunk 4]
```diff
@@ -1,113 +1,121 @@
...
                     children: ['Test Topic', 'Other Topic'],
                     entries: {
                         'Test Topic': {
-                            is_cohorted: is_cohorted,
+                            is_cohorted: true,
                             id: 'test_topic'
                         },
                         'Other Topic': {
-                            is_cohorted: is_cohorted,
+                            is_cohorted: true,
```
Comment (itsjeyd) [inline common/static/common/js/spec_helpers/discussion_spec_helper.js]: @pomegranited This seems to set `is_cohorted` to `true` unconditionally now. Is that intended?

**Comment by itsjeyd** [comment]
@pomegranited I reviewed the code and left some comments. It looks great overall, thanks for making an effort to reduce the diff between the master and solution forks! :)

> * I am unsure how to test the MathJax discussion functionality? @itsjeyd can you help?

Maybe just try entering some MathML when creating a post/response and see if you (a) get a correct preview and (b) whatever formula you entered is rendered correctly after you add the post/response. Not sure if there is any other functionality that needs to be tested w/r/t MathJax.

> * The code in [lms.django_comments_client.forum] still needs to be moved to lms.discussion. See edx:master changes starting from 89f93df.

Not completely sure if this is part of the scope of this task. If the "Discussion" tab is working as intended in the LMS after the changes introduced here, that should be enough to meet the requirements I think. To confirm with @bradenmacdonald.

---

It might not be necessary to make any changes based on my comments. But in case it is, I'll hold off on manual testing until you've had a chance to address them, to increase the chance of avoiding multiple cycles of manual testing.

File: common/static/common/js/spec_helpers/discussion_spec_helper.js (hunk 3 above)
Comment (pomegranited) [inline common/static/common/js/spec_helpers/discussion_spec_helper.js]: Yep, [upstream is the same](https://github.com/edx/edx-platform/blob/master/common/static/common/js/spec_helpers/discussion_spec_helper.js#L25-L27).

File: common/static/common/js/spec_helpers/discussion_spec_helper.js (hunk 4 above)
Comment (pomegranited) [inline common/static/common/js/spec_helpers/discussion_spec_helper.js]: I searched `edx-solutions:master` for calls to `makeCourseSettings`, and none of them passed an `is_cohorted` parameter.  The original code defaulted to `true`, so it looked fine to accept this upstream version.

File: common/static/common/js/discussion/views/discussion_content_view.js (hunk 1 above)
Comment (pomegranited) [inline common/static/common/js/discussion/views/discussion_content_view.js]: Ah good catch!  251f758 merges upstream's discussion/utils.js, and adds the 5th `beforeSend` parameter.

File: common/static/common/js/discussion/views/discussion_thread_profile_view.js (hunk 2 above)
Comment (pomegranited) [inline common/static/common/js/discussion/views/discussion_thread_profile_view.js]: It totally is, otherwise MathJax syntax isn't rendered on the user profile page!  Thanks for catching that.  Fixed with 89ce37f.

**Comment by pomegranited** [comment]
@itsjeyd Thanks for your comments!  I've addressed them inline.

* Re MathJax: have updated the description to provide an example of MathJax syntax that works.  Also pushed 4aa1e42, which uses the `DiscussionUtil.convertMath` function, bringing those files into line with upstream.
* Re `lms.django_comments_client.forum`, I agree this is out of scope for this task, but am unsure where it fits in the overall rebase.  Have updated the description to state this.

Code should be stable enough now for manual testing.

**Comment by itsjeyd** [comment]
@pomegranited I tested this according to the description (thanks for the nice instructions!), mainly focusing on the `staff` user (cf. below). It's looking good already; most features were working as expected. Here are the things I noticed:

* If there are two units belonging to a subsection, and one of them (let's call it "X") contains an instance of `DiscussionCourseXBlock`, no posts are displayed and it is not possible to create a new post when switching to "X" from the other unit. After reloading the page (which brings up "X" right away), threads show up and the "New Post" button works.

* When creating first post on "Discussion" tab: Had to refresh "Discussion" tab after pinning/unpinning post to get "Pinned" label to appear/disappear in box listing posts by category (following/unfollowing a post worked fine without refreshing page). *I only saw this once and wasn't able to reproduce it later on.* So there's probably no need to spend a lot of time investigating if you can't reproduce it on your end.

* I did not get a modal when trying to create posts with the LMS service stopped; messages were displayed inline instead. Screenshots:

  * `DiscussionXBlock` (layout looks a little weird): inline-discussion-block-error-message-with-lms-stopped

  * "Discussion" tab:  discussion-tab-error-message-with-lms-stopped

  * `DiscussionCourseXBlock` (layout looks a little weird): course-discussion-block-error-message-with-lms-stopped

I only tested voting/unvoting and reporting/unreporting with a non-staff user so far; will test the main features with a non-staff user once the items listed above have been sorted out.

**Comment by pomegranited** [comment]
@itsjeyd I believe I've addressed your comments:

* Fixed `DiscussionCourseXBlock` loading issues with 6e93bf9.  This change is also related to the initial JS errors we were seeing when loading the DiscussionCourseXBlock, by moving the logic for initializing the `DiscussionApp` object out of `discussion/main.js` and into the JS for the xblock, and for the Discussion tab.
* I was unable to reproduce the pin/unpin issue you saw.
* Fixed error panel style issues with 12f3cb9.  
    `DiscussionCourseXBlock` screenshot: screen shot 2016-09-21 at 3 10 07 pm

    Inline Discussion XBlock screenshot: screen shot 2016-09-21 at 3 13 43 pm

File: lms/static/sass/discussion/utilities/_shame.scss [hunk 5]
```diff
@@ -165,3 +165,14 @@ li[class*=forum-nav-thread-label-] {
     margin-bottom: 0 !important; // overrides "#seq_content h1" styling
   }
 }
+
+// -------
+// Inline Discussion and Discussion Course Overrides
+// -------
+.vert-mod>div {
```
Comment (itsjeyd) [inline lms/static/sass/discussion/utilities/_shame.scss]: @pomegranited Nit: Add spaces around `>`.

File: common/static/common/js/discussion/main.js [hunk 6]
```diff
@@ -34,13 +33,20 @@
                     course_settings: course_settings
                 });
                 /* jshint +W031*/
-                return Backbone.history.start({
-                    pushState: true,
-                    root: "/courses/" + $$course_id + "/discussion/forum/"
-                });
+
+                // Avoid re-initializing Backbone.history
+                if(!Backbone.History.started) {
```
Comment (itsjeyd) [inline common/static/common/js/discussion/main.js]: @pomegranited Nit: Add space after `if`.

File: lms/static/sass/discussion/utilities/_shame.scss (hunk 5 above)
Comment (itsjeyd) [inline lms/static/sass/discussion/utilities/_shame.scss]: @pomegranited Nit: Add spaces around `>`.

File: common/static/common/js/discussion/main.js (hunk 6 above)
Comment (itsjeyd) [inline common/static/common/js/discussion/main.js]: @pomegranited Nit: Add space after `if`.

**Comment by itsjeyd** [comment]
Thanks @pomegranited for the updates! The code looks good and fixes the issues that I mentioned in my previous [comment](https://github.com/edx-solutions/edx-platform/pull/743#issuecomment-247974346). Can't reproduce the pin/unpin issue either.

So this is pretty much ready to go, except:

> View Discussion User Profile pages by clicking on the user's name. Posts should render, including MathJax markup.

This worked when I did the first round of testing, but it didn't work for me today: Profile pages loaded successfully, but didn't show any posts. There were no errors in the console. Same results for `staff` and non-staff user (`audit`). Could you double-check on your end please?

**Comment by pomegranited** [comment]
@itsjeyd Sigh, sorry about that! Should be ok now.

* d4fb6f4 fixes the issue with the User Profile not loading forum posts.
* daaa4b9 moves the DiscussionApp initialization logic into a function, so it can be called from the discussion_course.js instead of having to duplicate it there.
* d029089 makes the User Profile page look nicer.

**Comment by itsjeyd** [comment]
@pomegranited No worries, there's a lot of moving parts here. Your updates check out; profile pages are listing threads again :) :+1:

- [x] I tested this: Works as advertised in PR description.
- [x] I read through the code
- [ ] ~~Includes documentation~~ This is a collection of bug fixes and refactorings, so no documentation updates necessary.

**Comment by antoviaque** [comment]
@ziafazal Would you be able to review this? @bradenmacdonald mentioned this would be for you : ) If possible early on Monday, to have a chance of getting it merged this sprint, as it's in the current commitments. Thanks!

**Comment by ziafazal** [comment]
@pomegranited I have followed the instructions and tested actions listed. Overall it LGTM 👍 . I have couple of observations. 
1) Few units test related to discussions are failing.
2) While testing with inline discussion I was getting a JS error attaching the screenshot. I don't know if its related to changeset in PR.
screen shot 2016-09-26 at 1 46 45 pm

**Comment by pomegranited** [comment]
Thanks for your review, @ziafazal !  

I'm not able to reproduce the JS error you reported?  Perhaps it's leftover static files from a previous branch?  I've had to run `git clean -fx` before to clear out all the old static files.

I also wasn't sure how much of the tests I needed to get fixed for this issue, since so many were broken.  Do you want me to work on them, or can I merge this PR?

**Comment by ziafazal** [comment]
@pomegranited we should fix at least python unit tests. CircleCI has failure on these 3 tests. I'm not sure if @itsjeyd is working on those or not. If we are going to fix those in separate PR then we good to merge this one.
screen shot 2016-09-26 at 2 52 13 pm

**Comment by pomegranited** [comment]
No worries, @ziafazal , thank you for pointing me to the specific failing tests.  

They are definitely related to this PR (see 01a9cf3), so I've submitted a fix (a595908) and will merge if it works.

**Comment by pomegranited** [comment]
@ziafazal CC @itsjeyd The discussion unit tests are passing again, so I'm going to squash and merge this PR.
